contracts/src/strealm.cairo

```diff
@@ -0,0 +1,294 @@
+        ) {
+            self
+                .emit(
+                    RewardTokenUpdated {
+                        old_address: self.StRealm_reward_token.read(),
+                        new_address: new_reward_token_address
+                    }
+                );
+            self.StRealm_reward_token.write(new_reward_token_address);
+        }
+
+        fn _update_reward_payer(
+            ref self: ComponentState<TContractState>, new_reward_payer_address: ContractAddress
+        ) {
+            self
+                .emit(
+                    RewardPayerUpdated {
+                        old_address: self.StRealm_reward_payer.read(),
+                        new_address: new_reward_payer_address
+                    }
+                );
+            self.StRealm_reward_payer.write(new_reward_payer_address);
+        }
+
+
+        fn _end_latest_flow(ref self: ComponentState<TContractState>) {
+            let flow_id = self.StRealm_latest_flow_id.read();
+            let mut flow: Flow = self.StRealm_flows.read(flow_id);
+            flow.end_at = starknet::get_block_timestamp();
+            self.StRealm_flows.write(flow_id, flow);
+        }
+
+        fn _start_new_flow(ref self: ComponentState<TContractState>, new_flow_rate: u256) {
+            let new_flow_id = self.StRealm_latest_flow_id.read() + 1;
+            let new_flow: Flow = Flow { rate: new_flow_rate, end_at: BoundedInt::max() };
+            self.StRealm_flows.write(new_flow_id, new_flow);
+        }
+
+
+        fn _reset_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            let mut stream: Stream = self.StRealm_streams.read(owner);
+            stream.start_at = starknet::get_block_timestamp();
+            stream.flow_id = self.StRealm_latest_flow_id.read();
+            self.StRealm_streams.write(owner, stream);
+        }
+
+        fn _end_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            if owner.is_non_zero() {
+                let mut stream: Stream = self.StRealm_streams.read(owner);
+                stream.start_at = 0;
+                stream.flow_id = 0;
+                self.StRealm_streams.write(owner, stream);
+            }
+        }
+
+
+        fn _streamed_amount(
+            ref self: ComponentState<TContractState>,
+            token_amount: u256,
+            stream_duration: u64,
+            flow_rate: u256
+        ) -> u256 {
+            token_amount * stream_duration.into() * flow_rate
+        }
+
+
+        /// must be calld before balance updates
+        fn _claim_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            if owner.is_non_zero() {
+                let stream: Stream = self.StRealm_streams.read(owner);
+                if stream.flow_id.is_non_zero() && stream.start_at.is_non_zero() {
+                    let flow: Flow = self.StRealm_flows.read(stream.flow_id);
```

Comment:
Can't it happen that there will be multiple unclaimed streams?

If I understand correctly, currently, only updating the flow rate increments the flow ID. But if this happens multiple times and the stream owner doesn't claim each time, they'll miss all the rewards in between.

Reply:
yes. Once a new flow rate is added, users have to manually claim what they've earned up until their stream ended before they can continue streaming. So in essence, adding a new flow rate stops all streams for all users. To continue streaming, users need to manually restart streaming.

There is no easy way to make flow continuous and seamless.  open to suggestions if you have any

Flow rate changes should be infrequent and subject to a lot of discussion so there is a lot of awareness

Reply:
I see. If it's going to be infrequent, then I guess it's not worth doing it, though I don't think it would be too complicated. Each flow would also have to track the start time and we'd have to track last claimed flow ID for each owner. Then it's just a matter of looping between the missing flow IDs, calculating the durations and rewards.

An alternative solution could be that we could have `reward_claim` take an owner argument and the DAO could subsidize the claiming of unclaimed rewards. Again, maybe not worth it plus we want active DAO members that should know when to claim, etc.

Resolving the convo, but happy to hear your thoughts.